chain/pool/src/lib.rs

```diff
@@ -30,46 +33,36 @@ impl TransactionPool {
         let signer_id = signed_transaction.transaction.signer_id.clone();
         let signer_public_key = signed_transaction.transaction.public_key.clone();
         self.transactions
+            .get_mut()
             .entry((signer_id, signer_public_key))
             .or_insert_with(Vec::new)
             .push(signed_transaction);
     }
 
+    /// Returns a draining structure that pulls transactions from the pool in the proper order.
+    /// It has an option to take transactions with the same key as the last one or with the new key.
+    /// When the iterator is dropped, the rest of the transactions remains in the pool.
+    pub fn draining_iterator(&mut self) -> DrainingIterator {
+        DrainingIterator::new(self)
+    }
+
     /// Take `min(self.len(), max_number_of_transactions)` transactions from the pool, in the
     /// appropriate order. We first take one transaction per key of (AccountId, PublicKey) with
     /// the lowest nonce, then we take the next transaction per key with the lowest nonce.
     pub fn prepare_transactions(
         &mut self,
         max_number_of_transactions: u32,
     ) -> Result<Vec<SignedTransaction>, Error> {
-        let mut sorted = false;
-        let mut result = vec![];
-        while result.len() < max_number_of_transactions as usize && !self.transactions.is_empty() {
-            let mut keys_to_remove = vec![];
-            for (key, txs) in self.transactions.iter_mut() {
-                if !sorted {
-                    // Sort by nonce in non-increasing order to pop from the end
-                    txs.sort_by_key(|a| std::cmp::Reverse(a.transaction.nonce));
-                }
-                let tx = txs.pop().expect("transaction groups shouldn't be empty");
-                if txs.is_empty() {
-                    keys_to_remove.push(key.clone());
-                }
-                result.push(tx);
-                if result.len() >= max_number_of_transactions as usize {
-                    break;
-                }
+        let mut res = vec![];
+        let mut iter = self.draining_iterator();
+        for _ in 0..max_number_of_transactions {
+            if let Some(tx) = iter.next(false) {
```

Comment:
Since iterator is used in the same place it is created we can implement `Iterator` for `DrainingIterator` and then do
```rust
self.draining_iterator().take(max_number_of_transactions).collect()
```
If it wasn't used in the same place we would've needed some lifetimes trickery.

Reply:
This goes away in part 3 as well (towards tests)